client/allocrunner/task_coordinator_test.go

```diff
@@ -0,0 +1,449 @@
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskAllowed(t, coord, initTask)
+	requireTaskAllowed(t, coord, sideTask)
+	requireTaskBlocked(t, coord, mainTask)
+}
+
+func TestTaskCoordinator_PoststartStartsAfterMain(t *testing.T) {
+	ci.Parallel(t)
+
+	logger := testlog.HCLogger(t)
+
+	alloc := mock.LifecycleAlloc()
+	tasks := alloc.Job.TaskGroups[0].Tasks
+
+	mainTask := tasks[0]
+	sideTask := tasks[1]
+	postTask := tasks[2]
+
+	// Make the the third task is a poststart hook
+	postTask.Lifecycle.Hook = structs.TaskLifecycleHookPoststart
+
+	shutdownCh := make(chan struct{})
+	defer close(shutdownCh)
+	coord := newTaskCoordinator(logger, tasks, shutdownCh)
+
+	// All tasks start blocked.
+	requireTaskBlocked(t, coord, sideTask)
+	requireTaskBlocked(t, coord, mainTask)
+	requireTaskBlocked(t, coord, postTask)
+
+	// Set initial state, prestart tasks are allowed to run, main and poststart
+	// are blocked.
+	states := map[string]*structs.TaskState{
+		sideTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		mainTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		postTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskAllowed(t, coord, sideTask)
+	requireTaskBlocked(t, coord, mainTask)
+	requireTaskBlocked(t, coord, postTask)
+
+	// Sidecar and main running, poststart allowed to run.
+	states = map[string]*structs.TaskState{
+		sideTask.Name: {
+			State:     structs.TaskStateRunning,
+			Failed:    false,
+			StartedAt: time.Now(),
+		},
+		mainTask.Name: {
+			State:     structs.TaskStateRunning,
+			Failed:    false,
+			StartedAt: time.Now(),
+		},
+		postTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskAllowed(t, coord, sideTask)
+	requireTaskAllowed(t, coord, mainTask)
+	requireTaskAllowed(t, coord, postTask)
+}
+
+func requireTaskBlocked(t *testing.T, c *taskCoordinator, task *structs.Task) {
+	ch := c.startConditionForTask(task)
+	requireChannelBlocking(t, ch, task.Name)
+}
+
+func requireTaskAllowed(t *testing.T, c *taskCoordinator, task *structs.Task) {
+	ch := c.startConditionForTask(task)
+	requireChannelPassing(t, ch, task.Name)
+}
+
+func requireChannelPassing(t *testing.T, ch <-chan struct{}, name string) {
+	testutil.WaitForResult(func() (bool, error) {
+		return !isChannelBlocking(ch), nil
+	}, func(_ error) {
+		t.Fatalf("%s channel was blocking, should be passing", name)
+	})
+}
+
+func requireChannelBlocking(t *testing.T, ch <-chan struct{}, name string) {
+	testutil.WaitForResult(func() (bool, error) {
+		return isChannelBlocking(ch), nil
+	}, func(_ error) {
+		t.Fatalf("%s channel was passing, should be blocking", name)
+	})
+}
+
+func isChannelBlocking(ch <-chan struct{}) bool {
+	select {
+	case <-ch:
+		return false
+	default:
+		return true
+	}
+}
```

Comment:
Since this is just a refactoring of existing code, I tried to keep the behaviour the same, including the tests, but the file rename created a new diff. Here's the old vs. new test diff: https://gist.github.com/lgfa29/6c7d7b6590a9ea377ad12e886c3bfd4c